File: algorithms/neetcode/algorithms/algorithms/src/main/java/com/mycompany/slidingwindow/BestTimeToBuyAndSellStock.java
package com.mycompany.slidingwindow;

public class BestTimeToBuyAndSellStock {

    public int maxProfit2(int[] prices) {
        int oldStockPrice = prices[0];
        int profit = 0;
        for(int i = 1; i<prices.length; i++){
            if(prices[i]<oldStockPrice){
                oldStockPrice = prices[i];
            }else{
                profit+=prices[i]-oldStockPrice;
                oldStockPrice = prices[i];
            }
        }
        return profit;
    }

    // not working there is bug need to fix.
    public int maxProfit(int[] prices) {
        int left = 0;
        int right = 1;
        int maxProfitResult = 0;
        while( left < prices.length) {
            if(prices[left] < prices[right]) {
                maxProfitResult = Math.max(maxProfitResult, prices[right] - prices[left]);
                right++;
            } else {
                left = right;
                right++;
            }
        }
        System.out.println(maxProfitResult);
        return maxProfitResult;
    }
    public static void main(String[] args) {
        BestTimeToBuyAndSellStock s = new BestTimeToBuyAndSellStock();
        int[] prices = {7,1,5,3,6,4};

        System.out.println(s.maxProfit2(prices));
    }
}

File: algorithms/neetcode/sliding_window/best-time-to-buy-and-sell-stock.py

from typing import List

class S:
    def maxProfit(self, prices: List[int]) -> int:
        l, r = 0, 1 # left=buy, right=sell
        maxP = 0

        while r < len(prices):
            # profitable ?
            if prices[l] < prices[r]:
                profit = prices[r] - prices[l]
                maxP = max(maxP, profit)
            else:
                l += 1
            r += 1

        return maxP

prices = [7,1,5,3,6,4]
print(S().maxProfit(prices))
